src/Evolution/Systems/GeneralizedHarmonic/Observe.hpp

```diff
@@ -0,0 +1,281 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <string>
+#include <tuple>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "Domain/Tags.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/Tags.hpp"
+#include "IO/Observer/Actions.hpp"
+#include "IO/Observer/ArrayComponentId.hpp"
+#include "IO/Observer/Helpers.hpp"
+#include "IO/Observer/ObservationId.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/ReductionActions.hpp"
+#include "IO/Observer/VolumeActions.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/Tags.hpp"
+#include "PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp"
+#include "PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "Time/Tags.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/MakeString.hpp"
+#include "Utilities/Numeric.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace GeneralizedHarmonic {
+namespace Actions {
+
+namespace observe_detail {}  // namespace observe_detail
+
+/*!
+ * \brief Temporary action for observing volume and reduction data
+ *
+ * A few notes:
+ * - Writes the solution and error in \f$\Psi_{ab}, \Pi_{ab}\f$, and
+ * \f$\Phi_{iab}\f$ to disk as volume data.
+ * - The RMS error of \f$\Psi\f$ and \f$\Pi\f$ are written to disk.
+ */
+struct Observe {
```

Comment:
Is it possible for this to use the generic observations infrastructure that was added a few weeks ago? The GrMhd executable currently uses it, and If it can be used for GH I think this file would then be unnecessary.

Reply:
This is a good point. The new generic observations infrastructure, that works through `EventsAndTriggers`, consists primarily of two observers: [dg::Events::ObserveErrorNorms](https://spectre-code.org/classdg_1_1events_1_1observeerrornorms_3_01volumedim_00_01tmpl_1_1list_3_01tensors_8_8_8_01_4_00_01eventregistrars_01_4) and [dg::Events::ObserveFields](https://spectre-code.org/classdg_1_1events_1_1observefields_3_01volumedim_00_01tmpl_1_1list_3_01tensors_8_8_8_01_4_00_01tc6487cc51554fb0bf44e7a56facd638b). It turns out that:

- `ObserveErrorNorms` can compute and write L2 norms for errors in user-provided variables against their *analytic* solution values. While GeneralizedHarmonic evolves `Psi, Phi, Pi`, the underlying KerrSchild solution does not provide any of these variables (it, in turn, only provides our usual 3+1 quantities). Therefore, we cannot use `ObserveErrorNorms` here to write out L2 norms of errors in `Psi, Phi, Pi`.
- `ObserveFields` can write out either (A) errors in tensors against their analytic solution counterparts, or (B) tensors directly. I can certainly use it here to write out constraint violations  in volume. Unfortunately, however, I cannot use `ObserveFields` to write out *errors* in `Psi, Phi, Pi` in volume (again, because these vars arent available in KerrSchild). 

I am currently observing:
- Reduction: errors against analytic solution for `Psi, Phi, Pi`, gauge and other constraints;
- Volume: errors against analytic solution for `Psi, Phi, Pi`, gauge and other constraints.
While both `ObserveErrorNorms` and `ObserveFields` are inapplicable to `Psi, Phi, Pi`, this means at best I can offload some of the current volume observation to `ObserveFields` while keeping others in `GeneralizedHarmonic::Actions::Observe`. That, however, is discouraged, as performing more than one volume observations at the same time can lead to unpredicted behavior (c.f. warning [here](https://spectre-code.org/classdg_1_1events_1_1observeerrornorms_3_01volumedim_00_01tmpl_1_1list_3_01tensors_8_8_8_01_4_00_01eventregistrars_01_4)).

Reply:
What is `ObserveFieldNorms` supposed to do?  Isn't `ObserveNorms` already enough?  
Also, I believe `ObserveFields` is already capable of (or easily extended to) observing any volume data

Reply:
Currently:
- `ObserveNorms` **cannot** write out norms of any constraints.
- `ObserveFields` can write out tensorial versions of constraints in volume, but again, not their norms in volume.

Reply:
Would writing a compute item that takes a pointwise norm be sufficient to allow the generic observers to be used instead of this?

Reply:
Pretty sure I already have this on a branch...will check and update this pr if so

Reply:
With the new option tag handling written by @kidder it should also be possible to use cache tags in compute items (to be confirmed by @kidder, may require a small extension of his PR). Once that works, I have the compute item coded up to put analytic solutions of fields into the DataBox. That will be necessary to feed into the pointwise norm compute item that @wthrowe mentioned.

Reply:
@wthrowe @geoffrey4444 @nilsleiffischer - yes, my plan was to switch to generic observers once we have #1633 . Sorry, I should have made this clear in the description here once #1633 was put up. Until that PR is merged, please consider the GH observer here as placeholder (i.e. which can be used to check the actual evolution, but need not be reviewed).

Reply:
this file has been removed - replaced by generic observers